src/step-function.ts

```diff
@@ -364,32 +362,20 @@ export namespace $SFN {
     return context.evalExpr(array, call, (_, { normalizeOutputToJsonPath }) => {
       const arrayPath = normalizeOutputToJsonPath().jsonPath;
 
-      const parameters = callbackfn.parameters.map((param, i) => {
-        const paramName = isIdentifier(param.name)
-          ? param.name.name
-          : // if the parameter is a binding pattern, we to assign the value to a variable first to access it in the map state.
-            context.newHeapVariableName();
-
-        return {
-          param: [
-            `${paramName}.$`,
-            i === 0
-              ? "$$.Map.Item.Value"
-              : i == 1
-              ? "$$.Map.Item.Index"
-              : arrayPath,
-          ],
-          states: isBindingPattern(param.name)
-            ? // if the param is a binding pattern, the value will be placed on the heap and then bound in the body
-              context.evalDecl(param, { jsonPath: `$.${paramName}` })
-            : undefined,
-        };
-      });
+      const [itemParam, indexParam, arrayParam] = callbackfn.parameters;
```

Comment:
I was recently looking at this code, do we properly handle the case where all of these parameteres could be `undefined`? Their types are all `ParameterDecl` instead of `ParameterDecl | undefined` - which is wrong.

https://stackoverflow.com/questions/50647399/typescript-accessing-an-array-element-does-not-account-for-the-possibility-of-u

Reply:
The functions that use them have been updated to handle undefined (these cases are well tested at least), but I'd like to try that flag and see how much of a mess it makes. Also don't like this implicit behavior by default.

```ts
  public evalParameterDeclForStateParameter(
    node: FunctionlessNode,
    ...parameters: [ParameterDecl | undefined, ASLGraph.JsonPath][]
```

Reply:
> Also don't like this implicit behavior by default.

Totally agree with this - it's semantically wrong.

Reply:
https://github.com/functionless/functionless/pull/398